lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseCoordinator.cs

```diff
@@ -0,0 +1,113 @@
+        /// </summary>
+        private long _shouldCloseTask = 0;
+
+        /// <summary>
+        /// Before the task is returned, this variable is set to 1.
+        /// Close handler will check this variable to decide if it needs to throw an exception.
+        /// </summary>
+        private long _isTaskStopped = 0;
+
+        /// <summary>
+        /// Waiting time for the task to close by itself
+        /// </summary>
+        private readonly int _enforceCloseTimeoutMilliseconds;
+
+        /// <summary>
+        /// An event that will wait in close handler until it is either signaled from Call method or timeout.
+        /// </summary>
+        private readonly ManualResetEventSlim _waitToCloseEvent = new ManualResetEventSlim(false);
+
+        /// <summary>
+        /// Handle task close event and manage the states, wait/signal when closing the task
+        /// </summary>
+        /// <param name="enforceCloseTimeoutMilliseconds">Timeout in milliseconds to enforce the task to close if receiving task close event</param>
+        [Inject]
+        private TaskCloseCoordinator([Parameter(typeof(EnforceCloseTimeoutMilliseconds))] int enforceCloseTimeoutMilliseconds)
+        {
+            _enforceCloseTimeoutMilliseconds = enforceCloseTimeoutMilliseconds;
+        }
+
+        /// <summary>
+        /// Handle Task close event.
+        /// Set _shouldCloseTask to 1 so that to inform the task to stop at the end of the current iteration.
+        /// Then waiting for the signal from Call method. Either it is signaled or after _enforceCloseTimeoutMilliseconds,
+        /// If the closed event is sent from driver, checks if the task has been stopped. If not, throw IMRUTaskSystemException 
+        /// to enforce the task to stop.
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        internal void HandleEvent(ICloseEvent closeEvent)
+        {
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
+            _waitToCloseEvent.Wait(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
+
+            if (closeEvent.Value.IsPresent())
+            {
+                var msg = Encoding.UTF8.GetString(closeEvent.Value.Value);
+                if (msg.Equals(TaskManager.CloseTaskByDriver))
+                {
+                    Logger.Log(Level.Info, "The task received close event with message: {0}.", msg);
+
+                    if (Interlocked.Read(ref _isTaskStopped) == 0)
+                    {
+                        throw new IMRUTaskSystemException(TaskManager.TaskKilledByDriver);
+                    }
+                }
+            }
+        }
+
+        internal bool ShouldCloseTask()
+        {
+            return Interlocked.Read(ref _shouldCloseTask) == 1;
+        }
+
+        internal void SetTaskStopped()
+        {
+            Interlocked.Exchange(ref _isTaskStopped, 1);
+        }
+
+        internal void Signal()
```

Comment:
I can combine SetTaskStopped and Signal as they are always called together. _isTaskStopped  is still needed because _waitToCloseEvent.Wait can get out by timeout as well, set is not the only way for it to get out.